Fix payone payone statuses processing. If invoice has been set to order, request is reject

diff --git a/Controller/TransactionStatus/Index.php b/Controller/TransactionStatus/Index.php
index aeac8f7..e4bda2e 100644
--- a/Controller/TransactionStatus/Index.php
+++ b/Controller/TransactionStatus/Index.php
@@ -164,10 +164,18 @@ class Index extends \Payone\Core\Controller\ExternalAction
         }

         $oOrder = $this->orderHelper->getOrderByTxid($this->getParam('txid'));
+
         if (!$oOrder) {
             return 'Order not found';
         }

+        $oOrder = $this->orderFactory->create()->loadByIncrementId($this->databaseHelper->getOrderIncrementIdByTxid($sTxid));
+        $invoiceCollectionSize = $oOrder->getInvoiceCollection()->getSize();
+
+        if ($this->getParam('txaction') == 'paid' &&  $invoiceCollectionSize > 0) {
+            return 'Order notified and has an invoice';
+        }
+
         if ($this->getParam('txaction') == 'appointed' && $oOrder->getStatus() == 'canceled') {
             // order was canceled in checkout, probably due to browser-back-button usage -> create a new order for incoming payment
             $oOrder = $this->substituteOrder->createSubstituteOrder($oOrder, false);
diff --git a/Helper/Order.php b/Helper/Order.php
index 3df83c0..4dab0be 100644
--- a/Helper/Order.php
+++ b/Helper/Order.php
@@ -99,11 +99,17 @@ class Order extends \Payone\Core\Helper\Base
      */
     public function getOrderByTxid($sTxid)
     {
+        if (!$sTxid) {
+            return;
+        }
+
         $sIncrementId = $this->databaseHelper->getOrderIncrementIdByTxid($sTxid);
         $oOrder = $this->orderFactory->create()->loadByIncrementId($sIncrementId);
+
         if ($oOrder && $oOrder->getId()) {
             return $oOrder;
         }
+
         return null;
     }

diff --git a/Model/ResourceModel/TransactionStatus.php b/Model/ResourceModel/TransactionStatus.php
index f8ad716..b543658 100644
--- a/Model/ResourceModel/TransactionStatus.php
+++ b/Model/ResourceModel/TransactionStatus.php
@@ -154,7 +154,7 @@ class TransactionStatus extends \Magento\Framework\Model\ResourceModel\Db\Abstra
                 'txaction' => $this->getParam('txaction'),
                 'sequencenumber' => $this->getParam('sequencenumber'),
                 'clearingtype' => $this->getParam('clearingtype'),
-                'txtime' => date('Y-m-d H:i:s', $this->getParam('txtime')),
+                'txtime' => date('Y-m-d H:i:s', (int) $this->getParam('txtime')),
                 'price' => $this->getParam('price'),
                 'balance' => $this->getParam('balance'),
                 'receivable' => $this->getParam('receivable'),
diff --git a/Setup/UpgradeSchema.php b/Setup/UpgradeSchema.php
index 057d931..a8a2b06 100644
--- a/Setup/UpgradeSchema.php
+++ b/Setup/UpgradeSchema.php
@@ -213,6 +213,13 @@ class UpgradeSchema extends BaseSchema implements UpgradeSchemaInterface
                 'receivable', ['type' => Table::TYPE_DECIMAL, 'length' => '20,4', 'default' => '0']
             );
         }
+
+        if (version_compare($context->getVersion(), '3.2.0.1', '<')) {
+            $setup->getConnection()->modifyColumn(
+                $setup->getTable(Api::TABLE_PROTOCOL_API),
+                'txid', ['type' => Table::TYPE_TEXT, 'length' => '32']
+            );
+        }
     }

     /**
diff --git a/etc/module.xml b/etc/module.xml
index 3b526fb..fc84f89 100644
--- a/etc/module.xml
+++ b/etc/module.xml
@@ -25,7 +25,7 @@
  */
 -->
 <config xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" xsi:noNamespaceSchemaLocation="../../../../../lib/internal/Magento/Framework/Module/etc/module.xsd">
-    <module name="Payone_Core" setup_version="3.2.0">
+    <module name="Payone_Core" setup_version="3.2.0.1">
         <sequence>
             <module name="Magento_Quote" />
             <module name="Magento_Sales" />
--
+2.30.1 (Apple Git-130)

